### DISCUSSION TITLE
# GSoC 2026 Proposal — Automated Agent System for Adding New Model Support to Optimum-Intel

### DISCUSSION THREAD

**Discussion by jaytiwarihub** [openvinotoolkit/openvino#35050] [Google Summer of Code]
**Organization:** OpenVINO Toolkit / Intel  
**Project:** #12  
**Mentors:** @rkazants, @popovaan 
**Applicant:** Jay Tiwari  
**GitHub:** @jaytiwarihub  
**Email:** jait66995@gmail.com
---

## 1. About Me

I'm a deep learning student with a strong interest in LLMs and model optimization.
Over the past year I've been contributing to open source ML tooling, mostly around
the OpenVINO ecosystem and HuggingFace.

My most relevant contributions:

**Keras OpenVINO backend ([keras-team/keras#22412](https://github.com/keras-team/keras/pull/22412))**
I implemented `nn.glu` (Gated Linear Unit) and `nn.sparsemax` for the Keras
OpenVINO backend. This wasn't just calling an existing function — I had to read
the mathematical definition of each operation, understand how OpenVINO's IR graph
works, and manually map each step to the right `ov_opset` nodes. I tested both
by modifying Keras source directly in Google Colab, removing the test exclusions,
and making the `NNOps` PyTest suite pass with `KERAS_BACKEND=openvino`.

**Keras Hub — Qwen 2.5 VL and Llama 3 ([keras-team/keras-hub#2574](https://github.com/keras-team/keras-hub/pull/2574), [#2552](https://github.com/keras-team/keras-hub/pull/2552))**
I ported the Qwen 2.5 Vision-Language Model to Keras Hub and added the
Llama-3-Groq-8B-Tool-Use preset. These were complex — multimodal projection
layers, large state dictionaries, non-standard configurations. I also maintain
the Llama 3 weights on Kaggle.

**TensorFlow-ONNX ([onnx/tensorflow-onnx#2425](https://github.com/onnx/tensorflow-onnx/pull/2425))**
Modernized test backend APIs for ONNX graph translation testing.

**Keras-IO (documentation and CI)**
Authored official documentation files and handled CI merge conflicts, which
taught me how strict ML repos are about contribution format.

I mention all of this not to list credentials but because it's directly relevant
to what this project is asking for — and I'll explain that connection below.

---

## 2. The Problem I Want to Solve

When I was reading through `optimum-intel` to understand this project, I opened
`model_patcher.py`. It's 8,321 lines long and has 40+ patcher classes. Every
single one of those classes was written by a human, by hand, for one specific
model family.

The process goes like this: a new model appears on HuggingFace. Someone on the
team has to study its `config.json`, figure out what's different about its
attention mechanism or positional encoding, write a config class in
`model_configs.py`, write a patcher class that fixes whatever `torch.jit.trace`
breaks on, wire it into the registry, create a tiny test model, write tests,
and update documentation.

The clearest illustration of this problem is
[PR #1626](https://github.com/huggingface/optimum-intel/pull/1626) — adding
support for `ai-sage/GigaChat3-10B-A1.8B-bf16`. A skilled, returning GSoC
contributor spent **38 commits over 4 weeks** on one model. Looking at the
commit history tells the story: week 1 figuring out the architecture and writing
the initial patcher, weeks 2-3 iteratively fixing test failures and merge
conflicts, week 4 rewriting the MoE routing into a vectorized form after reviewer
feedback. The majority of that time was spent on tasks that follow known,
repeatable patterns — not on genuinely novel problems.

At the same time, open issues
[#1628](https://github.com/huggingface/optimum-intel/issues/1628) (Qwen3.5),
[#1647](https://github.com/huggingface/optimum-intel/issues/1647) (Kimi-K2.5),
and [#1338](https://github.com/huggingface/optimum-intel/issues/1338) (Mistral3)
are all sitting open — each one waiting for a human to go through the same
process again. With 60+ new decoder-only LLMs released on HuggingFace every
month, the backlog never closes.

The project asks for an agent system that automates this. I think I can build it,
and I want to explain exactly how.

---

### The core idea

You give the tool a HuggingFace model ID:

```bash
python add_model.py --model_id "Qwen/Qwen3-8B"
```

It analyses the model, retrieves similar existing examples from the codebase,
generates the config and patcher code using a local LLM, runs the export to
check if it works, and gives you back a folder of ready-to-review files.

The output looks like this:

```
output/Qwen3-8B/
├── model_configs_patch.py     ← paste into model_configs.py
├── model_patcher_patch.py     ← paste into model_patcher.py
├── test_qwen3.py              ← pytest file
├── tiny_model/
│   └── config.json            ← reduced config for tests
└── REVIEW.md                  ← notes on what to check before submitting
```

### The core insight behind the design

Before explaining the nodes, it's worth stating the key insight that makes
this system practical rather than speculative.

Reading `model_patcher.py` directly — 8,321 lines, 40+ classes — reveals
something important: there are only 5 distinct patcher families. Base decoder,
MoE, sliding window, custom attention, and RoPE variants. Every new model maps
to one of these. Most new decoder-only LLMs inherit `OVDecoderModelPatcher`
with zero custom code at all.

This means the solution space is bounded and retrievable, not open-ended.
The agent doesn't guess — it retrieves the closest real working example from
the existing codebase and adapts it. That's why RAG over the real repo is the
foundation, not a nice-to-have.

### Why I chose LangGraph

The project description mentions LangGraph specifically, and it's genuinely the
right tool here. The pipeline has conditional branching — if the export fails,
I need to retry code generation with the error message attached. LangGraph's
`StateGraph` with conditional edges handles this cleanly without me having to
build a custom orchestration loop.

The full graph is six nodes:

<details>
<summary><strong>🖼️ Click to preview — Full Pipeline Overview</strong></summary>
<br>

![Full Pipeline Overview](https://github.com/user-attachments/assets/ffc4f3b4-4fd9-4a10-b863-dcfdcbf69261)

</details>

> **Note:** The pipeline is deterministic everywhere except Node 3 — minimising
> hallucination surface to only the parts that genuinely need language understanding.

I deliberately made as many nodes as possible deterministic (no LLM). The LLM
only runs where it actually needs to — for code generation. Everything else is
logic that doesn't need language understanding.

---

### Node 1 — Model Analyser

This reads `config.json` from HuggingFace and builds a structured profile of
the model:

```python
@dataclass
class ModelProfile:
    arch_type: str            # "decoder", "encoder-decoder", "vlm"
    attention_type: str       # "mha", "gqa", "mla", "swa"
    positional_encoding: str  # "rope", "alibi", "learned", "none"
    is_moe: bool
    trust_remote_code: bool
    hidden_size: int
    num_hidden_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    max_position_embeddings: int
```

**Fix — `trust_remote_code` blindspot:** Without this step, the `ModelProfile`
for models like Qwen, InternLM, and ChatGLM would be incomplete — their real
architecture lives in a custom `modeling_xyz.py` file, not in `config.json`.
Node 1 detects `trust_remote_code=True` and AST-parses that file to extract the
true attention classes, forward method signatures, and non-standard operations
before anything else runs.

Node 1 also performs a runtime module inspection fallback for truly novel
architectures:

```python
for name, module in pipe.named_modules():
    class_name = type(module).__name__
    if 'Attention' in class_name or 'MoE' in class_name:
        print(f"{name}: {class_name}")
```

This catches custom attention implementations that aren't described in either
`config.json` or the modeling file's class names.

---

### RAG Retrieval

Before generating any code, the system retrieves the 3 most similar existing
models from the optimum-intel codebase to use as few-shot examples.

I deliberately avoided heavy dependencies here. The corpus is ~40 config classes
and ~40 patcher classes — adding ChromaDB or Pinecone for 40 items would be
bad engineering. Instead:

**Level 1** — pure Python dict lookup. Match the new `ModelProfile` against
pre-indexed profiles by structural features (arch type, attention type, positional
encoding, is_moe). Returns top 5.

**Level 2** — numpy cosine similarity. Embed the source code of those 5
candidates using `all-MiniLM-L6-v2` (80MB, loaded once, cached in memory).
Re-rank by semantic similarity. Returns top 3.

These 3 examples go into the LLM prompt as few-shot templates.

**How the knowledge base is built — from git diffs, not just file source:**

Rather than embedding the current state of each file, the knowledge base is built
from structured git diffs of merged PRs. For each merged model-support PR:

```bash
git fetch origin pull/<PR_NUMBER>/head:pr-<PR_NUMBER>
git diff main...HEAD -- optimum/exporters/openvino/model_configs.py \
                        optimum/exporters/openvino/model_patcher.py
```

This captures what changed and why. PR #1626 (GigaChat3) shows exactly how the
DeepSeek patcher was extended with a vectorized MoE block and a three-way
interface fallback. That diff pattern becomes a retrieval target for any future
MoE model. The knowledge base grows automatically as new models merge.

**The "reuse existing config" detection:**

Not every new model needs a new config class. GigaChat3 reused
`DeepseekOpenVINOConfig` entirely — only the transformers version bounds needed
updating. If Level 1 retrieval returns a match with similarity > 0.90, Node 3
first checks whether the existing config class can be reused with version bound
updates only. Only if the architecture genuinely differs does it generate a new
config class.

<details>
<summary><strong>🖼️ Click to preview — RAG Two-Level Retrieval</strong></summary>
<br>

![RAG Two-Level Retrieval](https://github.com/user-attachments/assets/e9a6a433-581e-4700-a925-bfaa885ff8ee)

</details>

> **Note:** Two-level retrieval avoids heavy dependencies while ensuring the LLM
> always generates from real working examples, not from scratch.

---

### Node 3 — Code Generator

This is the LLM node. It receives the `ModelProfile` and the 3 retrieved examples,
and generates both files in a single call:

- The `model_configs.py` entry with the correct `@register_in_tasks_manager`
  decorator and task list, or a version-bound update to an existing config if
  reuse detection triggers
- The `model_patcher.py` entry inheriting from the correct patcher family

A few things I hardcode into the system prompt based on things I learned reading
the actual codebase and studying real PRs:

**KV cache constraint:** Without this rule, the LLM writes standard PyTorch cache
logic that breaks OpenVINO's stateful transformation in `stateful.py`. The system
prompt explicitly instructs the LLM to replace any custom cache class with standard
`DynamicCache` and leave the rest to `patch_stateful()` downstream.

**Vectorized MoE pattern:** For MoE models, the system prompt includes the
canonical vectorized routing pattern derived from
[PR #1626](https://github.com/huggingface/optimum-intel/pull/1626) (GigaChat3)
and [PR #1569](https://github.com/huggingface/optimum-intel/pull/1569) (Afmoe).
The key constraint: replace any Python for-loop over experts with a `torch.bmm`
based vectorized implementation. Non-vectorized MoE produces excessive graph
nodes in OpenVINO IR and the conditional `if mask.any()` branch causes incorrect
tracing. This is the pattern that took a contributor 3 weeks to arrive at through
reviewer feedback — the system knows it upfront.

**Three-way interface fallback for MoE:** The prompt handles three interface
variants seen in the wild: `moe_infer` (older transformers), `moe` method
(newer transformers >= 4.57), and a fallback for models exposing `experts`
directly. This was the exact pattern @rkazants requested in his review of PR #1626.

**Version-conditional patches go to REVIEW.md:** If a patch requires knowing a
specific OpenVINO or transformers version boundary, the LLM writes a flag comment
and adds it to `state.flags`. These are written to `REVIEW.md` for human review.

**Local LLM by default:** I use LiteLLM as an abstraction layer so the tool
works with any backend. The default is `Qwen2.5-Coder-7B-Instruct` quantized
to INT4, running locally via `openvino-genai`.

```python
# Default — local OpenVINO GenAI
llm = LiteLLM(model="openvino/Qwen2.5-Coder-7B-Instruct-INT4")

# Optional fallback
llm = LiteLLM(model="openai/gpt-4o")
```

---

### Node 4 — Tiny Model Generator + Test Writer

For testing, I only shrink two things:

```python
def make_tiny_config(profile: ModelProfile) -> dict:
    tiny = profile.original_config.copy()
    tiny["num_hidden_layers"] = 2
    tiny["max_position_embeddings"] = 64
    # hidden_size, num_attention_heads, num_key_value_heads,
    # intermediate_size — all untouched
    return tiny
```

The test writer generates updates across all five required test files:

- `tests/openvino/test_decoder.py` — adds model to `SUPPORTED_ARCHITECTURES`,
  sets expected SDPA count, adds any model-specific tokenizer quirks
- `tests/openvino/test_export.py` — adds export test entry
- `tests/openvino/test_exporters_cli.py` — adds CLI test entry
- `tests/openvino/test_quantization.py` — adds model to
  `SUPPORTED_ARCHITECTURES_WITH_AUTO_COMPRESSION`
- `tests/openvino/utils_tests.py` — adds model ID mapping

---

### Node 5 — Registry Wiring

Without this step, the generated patcher class exists in the file but
Optimum-Intel silently never loads it. Node 5 deterministically inserts:

- Architecture string into `TasksManager` registry
- Patcher mapping inside the config class's `patch` method
- Entry in `docs/source/openvino/models.mdx` with first letter capitalised

No LLM, no hallucination risk.

---

### Node 6 — Validator and Self-Correction

This node actually runs the export. If it fails:

1. Parses the traceback — `TracingError` → patcher problem, `KeyError` → config problem
2. Re-prompts Node 3 with: original prompt + failed code + error message
3. Retries up to 3 times

If after 3 retries it still fails, or if the model needs version-conditional
patches, it writes a `REVIEW.md` explaining exactly what a human needs to check.

### LangGraph state design

```python
class AgentState(TypedDict):
    model_id: str
    profile: ModelProfile
    examples: list[str]
    config_code: str
    patcher_code: str
    test_code: str
    error: str | None
    retries: int
    flags: list[str]  # written to REVIEW.md
```

```python
def should_retry(state: AgentState) -> str:
    if state["error"] and state["retries"] < 3:
        return "generate_code"
    elif state["error"]:
        return "flag_for_review"
    else:
        return "end"
```

<details>
<summary><strong>🖼️ Click to preview — LangGraph StateGraph</strong></summary>
<br>

![LangGraph StateGraph](https://github.com/user-attachments/assets/81c485e6-0ea1-454a-bc74-80a5991bfe19)

</details>

> **Note:** The conditional retry edge mirrors the real review process — the
> same class of errors that took 38 commits in PR #1626, now automated.

---

### Why my background is relevant here

When I implemented `nn.glu` and `nn.sparsemax` for the Keras OpenVINO backend,
I was doing the same thing this agent needs to do — reading a spec, understanding
it, and writing OpenVINO-compatible code by hand. The cognitive loop is identical.

I've felt the manual pain this project addresses firsthand. I know where it breaks.

---

## 4. Scope

**Phase 1 (the whole summer):** Decoder-only causal LLMs only.

**Success gate:** 70%+ success rate on a benchmark of 10 known decoder-only
models. Success means: export completes, IR loads, forward pass produces correct
output shapes.

**Phase 2 (only if Phase 1 gate is met):** Extend to VLMs.

Encoder-decoder models, diffusion models, and audio models are out of scope.

---

## 5. Timeline

| Week | What I'm doing | What I'll have |
|------|----------------|----------------|
| 1 | Deep-read `model_configs.py`, `model_patcher.py`, study PR #1626 and PR #1569 | Architecture map |
| 2 | Build knowledge base from git diffs of merged PRs | Searchable feature index |
| 3 | Node 1 — Model Analyser + AST parser + runtime inspection | Working analyser |
| 4 | RAG retrieval — Level 1 + Level 2 | Working retrieval |
| 5 | Node 3 — Config Writer with reuse detection | Config generation |
| 6 | Node 3 — Patcher Writer with MoE vectorization | **Mid-term: full pipeline on Llama-3-8B** |
| 7 | Edge cases — MoE, sliding window, custom RoPE | Robust generation |
| 8 | Node 4 five-file test writer + Node 5 Registry Wiring | Deterministic nodes |
| 9 | Node 6 — Validator + self-correction loop | Full pipeline |
| 10 | Benchmark on 10 models, hit 70% gate | Benchmark results |
| 11 | Phase 2 VLM or Phase 1 hardening | Extension or robustness |
| 12 | Documentation, cleanup, PR | Final deliverable |

---

## 6. Why I Think This Is Doable

Reading `model_patcher.py` directly revealed only 5 distinct patcher patterns
across 8,321 lines. Most new decoder-only LLMs inherit `OVDecoderModelPatcher`
with no custom code at all.

[PR #1626](https://github.com/huggingface/optimum-intel/pull/1626) is the clearest
example of the problem — 38 commits over 4 weeks. The MoE vectorization pattern
that took 3 weeks through reviewer feedback is hardcoded into the system prompt
from day one. The parts that genuinely require human judgment — OpenVINO
version-specific bug workarounds, transformers version bounds — are flagged
honestly in `REVIEW.md` rather than guessed at.

---

## 7. References

- `huggingface/optimum-intel` — target repository
- [PR #1626](https://github.com/huggingface/optimum-intel/pull/1626) — GigaChat3 support
- [PR #1569](https://github.com/huggingface/optimum-intel/pull/1569) — Afmoe MoE reference
- [Issue #1628](https://github.com/huggingface/optimum-intel/issues/1628) — Qwen3.5
- [Issue #1647](https://github.com/huggingface/optimum-intel/issues/1647) — Kimi-K2.5
- [Issue #1338](https://github.com/huggingface/optimum-intel/issues/1338) — Mistral3
- [keras-team/keras#22412](https://github.com/keras-team/keras/pull/22412) — OpenVINO GLU and Sparsemax
- [keras-team/keras-hub#2574](https://github.com/keras-team/keras-hub/pull/2574) — Qwen 2.5 VL
- [keras-team/keras-hub#2552](https://github.com/keras-team/keras-hub/pull/2552) — Llama 3
- [onnx/tensorflow-onnx#2425](https://github.com/onnx/tensorflow-onnx/pull/2425) — backend tests
- LangGraph, LiteLLM, OpenVINO GenAI

---

*I wrote this proposal myself. Every technical decision in here is something
I can explain and defend — including the ones where I'm not sure yet.*
````
